### DISCUSSION TITLE
AWS SAM and Typescript: building functions and layers

### DISCUSSION THREAD

**Discussion by daniele-salvagni** [daniele-salvagni/dan.salvagni.io#7] [Comments]
**Comments for the blog post:** https://dsalvagni.com/b/aws-sam-and-typescript-building-functions-and-layers/

**Comment by samputer** [comment]
Fantastic. Just what I needed - thank you!

Note that I needed to copy any `node_modules` included in the layer across also.

**Reply by daniele-salvagni** [reply]
Thank you for bringing this up. You are absolutely right that if the layer has dependencies, they'll also need to be copied over during the build process. I will update this post at some point to also cover this case.

**Reply by danhnguyeen** [reply]
Hi @daniele-salvagni , 
thanks for your great sharing. I have tried your solution and everything worked well. 
But due to we copy everything in `node_module`, the size of layer will be very large.
I tried to download a very simple layer and extract the zip file, and the layer size is 71MB (because `node_module` includes `typescript` package also). So some of my Lambda functions threw bellow error
`Function code combined with layers exceeds the maximum allowed size of 262144000 bytes`
It means my lambda function + layer is large than the limitation 250MB even my project is still very simple. 
Do you have any idea about this case?

**Reply by danhnguyeen** [reply]
OK, I know why. In your example, I think we should move `typescript` to `devDependencies` instead `dependencies`. It will reduce the package size a lot :)

**Reply by daniele-salvagni** [reply]
Thanks for catching that mistake @danhnguyeen. Yes, `typescript` and `@types/node` should both have been inside `devDependencies`. I will fix the examples asap

**Comment by gittemo** [comment]
very useful. Could you be so kind and put up a GH repo with this?

**Reply by hafizali05** [reply]
@giscus this would be greatly appreciated

**Reply by daniele-salvagni** [reply]
Hi, I just created a repository on GitHub with a full working example:
https://github.com/daniele-salvagni/aws-sam-ts-functions-layers

**Comment by hafizali05** [comment]
When in lambda trying to import module from layer, gives module not sound error, can you please suggest ?

**Reply by daniele-salvagni** [reply]
If the error is local (in your editor) you could create a global `package.json` where you put the following inside devDependencies so that you editor knows where to look for the module:

```
"commons": "file:./layers/commons"
```

You can reference [this repo]( https://github.com/daniele-salvagni/aws-sam-ts-functions-layers) I just added for a full example

**Comment by gittemo** [comment]
Thank you so much for this ☺️Am 22.05.23, 16:42 schrieb Daniele Salvagni ***@***.***>:

   
  Hi, I just created a repository on GitHub with a full working example: https://github.com/daniele-salvagni/aws-sam-ts-functions-layers

**Comment by ahmednmourad** [comment]
When you try to copy "commons" folder into "node_modules" there might be a package with the same name, so I would suggest naming "commons" package differently to avoid conflicts.

**Comment by sumitbusa** [comment]
Hi, 

just one question, lets assume i have two files in hello_world functions, instead of only one app.ts.

what will be change, based on current code base i believe it will only compile app.ts instead all present.

**Reply by daniele-salvagni** [reply]
If the other file is some code imported in `app.ts`, esbuild will bundle it inside `app.js` when transpiling it to Javascript, so you don't need to make any changes.

If you want to include other unrelated files inside your lambda, you will need to use a custom Makefile (like I did for the layer) instead of using the provided esbuild integration. Inside your Makefile you could decide to transpile the code with either `esbuild` or `tsc`, copy over any other files you need, or run any other command you need to bundle your lambda.

**Comment by kostekd** [comment]
Fantastic article. Just a little update I believe currently it is possible to specific a format option for esbuild in SAM AWS. 

Link to the newest documentation: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-using-build-typescript.html

**Reply by daniele-salvagni** [reply]
Thanks for the heads up, happy to hear that they finally exposed the esbuild options!

**Reply by kostekd** [reply]
Since reading your article, I've been experimenting with AWS SAM, Lambda layers, and functions. Your article provided a great starting point!

Currently, my entire repository is natively written in TypeScript with ES modules. I'm using Rollup to transpile and bundle my layer.

Maybe in the future, I'll find time to write an updated version of an article like yours!
